### ISSUE TITLE
Pyright should not allow assigning to a method

### ISSUE THREAD

**Issue by LeeeeT** [microsoft/pyright#9426]
You're allowed to override a method in a child class with a new method whose type is a subtype of the parent's method type. Therefore, method types should be covariant and therefore you can't assign to them, but pyright allows you to assign to a method. Here's an example of how this false negative can hide a bug:
```python
class A:
    def f(self, v: str) -> str: ...

class B(A):
    def f(self, v: object) -> str: ...

def mutate(cls: type[A]):
    cls.f = lambda self, v: v.lower()  # pyright is ok with this

mutate(B)

B().f(1)  # AttributeError
```

Mypy disallows assigning to a method.

**Comment by erictraut** [comment]
Pyright doesn't currently check for this case, but I don't consider this a bug. I'll treat this as an enhancement request and consider it in that light.

I don't think this is a common source of bugs, and it will produce false positives in some code bases, so I'm not sure entirely convinced this feature is worth it. It would probably require a dedicated diagnostic rule (e.g. `reportAssignMethod`) so it can be disabled independently from other diagnostic checks.

As you mentioned, mypy supports a check ("method-assign") that detects assignments to methods. It doesn't appear to distinguish between method assignments that are type safe and those that are not. For example, in the above example, if `mutate` were modified to assign to `A.f` rather than `cls.f`, it would be type safe. I'm not sure if this is just an oversight in the mypy implementation or if it was just done for expediency.

I'll need to give this more thought.

**Comment by LeeeeT** [comment]
> if `mutate` were modified to assign to `A.f` rather than `cls.f`, it would be type safe. I'm not sure if this is just an oversight in the mypy implementation or if it was just done for expediency.

The same argumentation can be used for the following case:
```python
def f(v: object):
    v.lower()
```
This function is type safe in some cases (when `v` happens to have type `str`). Should a type checker try to distinguish the usages that are safe and those that are not? Of course, not. When you have a value of type `type[A]`, a type checker should check that your code is type safe for any values of that type (including `B`) and generate an error if it doesn't work for some of them.

Please tell me if I misunderstand what a type checker should do and what's the difference between the example above and that in my first comment.

**Comment by erictraut** [comment]
A type checker provides a set of tools to help developers write robust (bug-free and maintainable) code. It is always possible, especially in a language like Python, to subvert a type checker. There are countless ways to do so, and it's not the job of a type checker to prevent you from doing all of these. 

A type checker necessarily makes tradeoffs. A check that is too lenient will allow common bugs to go undetected. A check that is too strict will catch very few *actual* bugs but will be a source of noise for developers. A good type checker strikes the right balance between these extremes.

The difference between your example above and your first comment is that it's quite common to mistakenly attempt to call a non-existent method on an object, and a type checker can detect this situation with a relatively low chance of a false positive. In contrast, it's rare for Python code to mutate a class by overwriting a method in the manner you've done above. It's even rarer for this to result in a runtime crash.

I'm curious whether this issue stems from a real-world example in your code or whether this was just a theoretical problem that you noticed? If it was a real-world example in your code, why was the code originally written such that it overwrites methods in some classes? This seems like a really unusual (and dangerous) thing to do in general.

**Comment by LeeeeT** [comment]
Thanks for the explanation! I found this false negative accidentally while playing around with the code from my previous reports. I've never used anything like that in real-world projects. I agree it might not be worth it to implement this enhancement request if it will result in false positives in other more common use cases or if it doesn't align well with your vision of pyright.

**Comment by sfllaw** [comment]
Note that mypy implemented the [method-assign](https://mypy.readthedocs.io/en/stable/error_code_list.html#check-that-assignment-target-is-not-a-method-method-assign) error because it cannot distinguish between bound and unbound callables. See https://github.com/python/mypy/issues/2427 for the discussion.

> ### Check that assignment target is not a method [method-assign]
> In general, assigning to a method on class object or instance (a.k.a. monkey-patching) is ambiguous in terms of types, since Python’s static type system cannot express the difference between bound and unbound callable types. Consider this example:
> 
> ``` python
> class A:
>     def f(self) -> None: pass
>     def g(self) -> None: pass
> 
> def h(self: A) -> None: pass
> 
> A.f = h  # Type of h is Callable[[A], None]
> A().f()  # This works
> A.f = A().g  # Type of A().g is Callable[[], None]
> A().f()  # ...but this also works at runtime
> ```
>
> To prevent the ambiguity, mypy will flag both assignments by default. If this error code is disabled, mypy will treat the assigned value in all method assignments as unbound, so only the second assignment will still generate an error.
